### PR TITLE
(torch.distributed.elastic) properly format traceback on error

### LINKED ISSUE #60910: [elastic launcher] restart on startup failure makes no sense
## Motivation

If a program failed to start, elastic shouldn't try to restart. It should remain down. There needs to be some reasonable logic that restarts only when the program has started normally and then failed after some epochs / steps.

Since some programs may take as long as 10-15 min to start (huge models on multi-node envs), I propose a new configurable timer CLI arg, which I'd say should default to at least 15min - based on my current experiments. e.g. `--restart_threhold`

So the restart feature should kick in only if that threshold has been passed and otherwise do nothing.

thank you!

### PR DESCRIPTION
**Pull request by kiukchung**
Summary:
Fixes a bug introduced in https://github.com/pytorch/pytorch/pull/64036 where the traceback of the error handler is printed out rather than the traceback of the actual exception.

Fixes https://github.com/pytorch/pytorch/issues/60910
Closes https://github.com/pytorch/pytorch/issues/60910

BEFORE (note that the `py_callstack` is NOT the traceback of the RuntimeError):
```
**************************************************************************************************************************************************************************************************************************************************
                                                                                                              run_script_path FAILED
==================================================================================================================================================================================================================================================
Root Cause:
[0]:
  time: 2021-09-14_22:01:06
  rank: 0 (local_rank: 0)
  exitcode: 1 (pid: 1092727)
  error_file: /tmp/torchelastic_aeyvjbpe/none_8zuih7tj/attempt_0/0/error.json
  msg:
    {
      "message": "RuntimeError: rasing error since --throw was specified",
      "extraInfo": {
        "py_callstack": [
          "  File \"<string>\", line 1, in <module>\n",
          "  File \"/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/spawn.py\", line 116, in spawn_main\n    exitcode = _main(fd, parent_sentinel)\n",
          "  File \"/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/spawn.py\", line 129, in _main\n    return self._bootstrap(parent_sentinel)\n",
          "  File \"/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/process.py\", line 315, in _bootstrap\n    self.run()\n",
          "  File \"/usr/local/fbcode/platform009/lib/python3.8/multiprocessing/process.py\", line 108, in run\n    self._target(*self._args, **self._kwargs)\n",
          "  File \"/data/users/kiuk/fbsource/fbcode/buck-out/dev/gen/caffe2/run#link-tree/torch/multiprocessing/spawn.py\", line 59, in _wrap\n    fn(i, *args)\n",
          "  File \"/data/users/kiuk/fbsource/fbcode/buck-out/dev/gen/caffe2/run#link-tree/torch/distributed/elastic/multiprocessing/api.py\", line 382, in _wrap\n    ret = record(fn)(*args_)\n",
          "  File \"/data/users/kiuk/fbsource/fbcode/buck-out/dev/gen/caffe2/run#link-tree/torch/distributed/elastic/multiprocessing/errors/__init__.py\", line 373, in wrapper\n    error_handler.record_exception(e)\n",
          "  File \"/data/users/kiuk/fbsource/fbcode/buck-out/dev/gen/caffe2/run#link-tree/torch/distributed/elastic/multiprocessing/errors/error_handler.py\", line 86, in record_exception\n    _write_error(e, self._get_error_file_path())\n",
          "  File \"/data/users/kiuk/fbsource/fbcode/buck-out/dev/gen/caffe2/run#link-tree/torch/distributed/elastic/multiprocessing/errors/error_handler.py\", line 26, in _write_error\n    \"py_callstack\": traceback.format_stack(),\n"
        ],
        "timestamp": "1631682066"
      }
    }

==================================================================================================================================================================================================================================================
Other Failures:
  <NO_OTHER_FAILURES>
**************************************************************************************************************************************************************************************************************************************************
```

AFTER (note the traceback is the traceback of the RuntimeError):
```
********************************************************************************
                             run_script_path FAILED
================================================================================
Root Cause:
[0]:
  time: 2021-09-14_21:49:25
  rank: 0 (local_rank: 0)
  exitcode: 1 (pid: 1014681)
  error_file: /tmp/torchelastic_q0zods2c/none_qwmz5dgj/attempt_0/0/error.json
  msg: Traceback (most recent call last):
    File "/data/users/kiuk/fbsource/fbcode/buck-out/dev/gen/caffe2/run#link-tree/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 361, in wrapper
      return f(*args, **kwargs)
    File "/data/users/kiuk/fbsource/fbcode/buck-out/dev/gen/caffe2/run#link-tree/torch/distributed/run.py", line 671, in run_script_path
      runpy.run_path(sys.argv[0], run_name="__main__")
    File "/usr/local/fbcode/platform009/lib/python3.8/runpy.py", line 265, in run_path
      return _run_module_code(code, init_globals, run_name,
    File "/usr/local/fbcode/platform009/lib/python3.8/runpy.py", line 97, in _run_module_code
      _run_code(code, mod_globals, init_globals,
    File "/usr/local/fbcode/platform009/lib/python3.8/runpy.py", line 87, in _run_code
      exec(code, run_globals)
    File "/home/kiuk/tmp/test.py", line 55, in <module>
      main()
    File "/data/users/kiuk/fbsource/fbcode/buck-out/dev/gen/caffe2/run#link-tree/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 361, in wrapper
      return f(*args, **kwargs)
    File "/home/kiuk/tmp/test.py", line 25, in main
      raise RuntimeError("rasing error since --throw was specified")
  RuntimeError: rasing error since --throw was specified

================================================================================
Other Failures:
  <NO_OTHER_FAILURES>
********************************************************************************
```

Test Plan:
(see summary for before and after)

`test.py` contents:
```
import argparse
import os
import sys

import torch
import torch.distributed as dist
import torch.nn.functional as F

from torch.distributed.elastic.multiprocessing.errors import record

def parse_args(argv):
    parser = argparse.ArgumentParser(description="test script")
    parser.add_argument("--init_method", type=str, default="env://")
    parser.add_argument("--backend", type=str, default="gloo")
    parser.add_argument("--throw", action="store_true", default=False)
    parser.add_argument("--exit", action="store_true", default=False)
    return parser.parse_args()

record
def main():
    args = parse_args(sys.argv[1:])

    if args.throw:
        raise RuntimeError("rasing error since --throw was specified")

    if args.exit:
        sys.exit(1)

    init_method=args.init_method
    backend=args.backend

    world_size = int(os.environ["WORLD_SIZE"])
    rank = int(os.environ["RANK"])

    print(f"initializing `{backend}` process group with rank={rank}, world_size={world_size} at {init_method}")

    dist.init_process_group(
        backend=backend,
        init_method=init_method,
        world_size=world_size,
        rank=rank)

    print(f"successfully initialized process group with rank={dist.get_rank()}, world_size={dist.get_world_size()}")

    t = F.one_hot(torch.tensor(rank), num_classes=world_size)
    dist.all_reduce(t)
    derived_world_size = torch.sum(t).item()
    if derived_world_size != world_size:
        raise RuntimeError(f"derived world size: {derived_world_size} != actual world size: {world_size}")
    else:
        print(f"sucessfully derived world size: {derived_world_size} (expected: {world_size}). Exiting")

if __name__ == "__main__":
    main()
```

run it as:

```
$ python -m torch.distributed.run --nproc_per_node 2 test.py --throw
```

Differential Revision: D30953731

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @cbalioglu @gcramer23

### CODE REVIEW

**Comment by kiukchung** [comment]
cc) @stas00 this oughta fix both presenting the correct traceback and the formatting of the traceback. Hopefully this is a better UX since all the tracebacks for all the failures are presented to the user as the last output. I also fixed the uber-wide "*****" border (constrained it to 80 char width so that the borders do not wrap in a reasonably sized terminal window).

For really large jobs where we observe the same exception signature over and over again, I think the next improvement would be to work on an exception aggregator/deduper that will only print "unique" error stack traces.

P.S. while I didn't add `hostname` to the printout the `rank` (global rank) information is there for the time being.

**Review by cbalioglu — APPROVED** [review]

**Comment by stas00** [comment]
> cc) @stas00 this oughta fix both presenting the correct traceback and the formatting of the traceback. Hopefully this is a better UX since all the tracebacks for all the failures are presented to the user as the last output. I also fixed the uber-wide "*****" border (constrained it to 80 char width so that the borders do not wrap in a reasonably sized terminal window).
> 
> For really large jobs where we observe the same exception signature over and over again, I think the next improvement would be to work on an exception aggregator/deduper that will only print "unique" error stack traces.
> 
> P.S. while I didn't add `hostname` to the printout the `rank` (global rank) information is there for the time being. Also not sure if this will make it to 1.9.1. If not it'll definitely be there for 1.10. Thanks!

Are the last 2 items on your TODO list? Or should an Issue be created so they aren't getting lost?

And given the imminence of 1.10, it'd be amazing to have those worked on soon if there are resources to do so. e.g. attaching a hostname to identify the node the root cause exception came from is a burning need. To remind this is to deal with a situation when some gpu goes haywire and must be excluded from the subsequent SLURM batching, but currently we can't tell which node to exclude. i.e. the issue is not the software, but the hardware.

**Comment by stas00** [comment]
@kiukchung, should the other dumps to the console be also nicely unwrapped in the same way the Root Cause dump does as in this PR?

with 1.11-rc1 I get:

```
Traceback (most recent call last):
  File "/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/run.py", line 728, in <module>
ERROR:torch.distributed.elastic.multiprocessing.errors.error_handler:{
  "message": {
    "message": "SignalException: Process 75531 got signal: 15",
    "extraInfo": {
      "py_callstack": "Traceback (most recent call last):\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/agent/server/api.py\", line 906, in _exit_barrier\n    store_util.barrier(\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/utils/store.py\", line 67, in barrier\n    synchronize(store, data, rank, world_size, key_prefix, barrier_timeout)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/utils/store.py\", line 53, in synchronize\n    agent_data = get_all(store, key_prefix, world_size)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/utils/store.py\", line 31, in get_all\n    data = store.get(f\"{prefix}{idx}\")\nRuntimeError: Socket Timeout\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py\", line 345, in wrapper\n    return f(*args, **kwargs)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/run.py\", line 724, in main\n    run(args)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/run.py\", line 715, in run\n    elastic_launch(\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/launcher/api.py\", line 131, in __call__\n    return launch_agent(self._config, self._entrypoint, list(args))\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/launcher/api.py\", line 236, in launch_agent\n    result = agent.run()\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/metrics/api.py\", line 125, in wrapper\n    result = f(*args, **kwargs)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/agent/server/api.py\", line 709, in run\n    result = self._invoke_run(role)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/agent/server/api.py\", line 877, in _invoke_run\n    self._exit_barrier()\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/agent/server/api.py\", line 906, in _exit_barrier\n    store_util.barrier(\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/multiprocessing/api.py\", line 60, in _terminate_process_handler\n    raise SignalException(f\"Process {os.getpid()} got signal: {sigval}\", sigval=sigval)\ntorch.distributed.elastic.multiprocessing.api.SignalException: Process 75531 got signal: 15\n",
      "timestamp": "1645830602"
    }
  }
}
```

this is a pointless noise that can't be parsed by a human - and multiplied by 384 times in my case.

So it either should be not printed at all, or formatted nicely so that a human can actually read it. I'd say it'd be great to have a flag to suppress it which would help a lot in the case of huge cluster training.

Does it make sense?

The root cause one is perfect:

```
Root Cause (first observed failure):
[0]:
  time      : 2022-02-26_00:04:37
  host      : jean-zay-iam47-ib0
  rank      : 368 (local_rank: 0)
  exitcode  : 1 (pid: 75185)
  error_file: /tmp/torchelastic__xe3l9jx/none_0hh_k6t8/attempt_0/0/error.json
  traceback : Traceback (most recent call last):
    File "/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 345, in wrapper
      return f(*args, **kwargs)
    File "/gpfswork/rech/six/commun/code/tr8b-104B/Megatron-DeepSpeed/pretrain_gpt.py", line 241, in main
      pretrain(train_valid_test_datasets_provider, model_provider, forward_step,
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/Megatron-DeepSpeed/megatron/training.py", line 187, in pretrain
      iteration = train(forward_step_func,
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/Megatron-DeepSpeed/megatron/training.py", line 846, in train
      train_step(forward_step_func,
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/Megatron-DeepSpeed/megatron/training.py", line 436, in train_step
      loss = model[0].train_batch(data_iter=data_iterator)
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/DeepSpeed-master/deepspeed/runtime/pipe/engine.py", line 329, in train_batch
      self._exec_schedule(sched)
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/DeepSpeed-master/deepspeed/runtime/pipe/engine.py", line 1338, in _exec_schedule
      self._exec_instr(**cmd.kwargs)
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/DeepSpeed-master/deepspeed/runtime/pipe/engine.py", line 691, in _exec_backward_pass
      super().backward(self.loss)
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/DeepSpeed-master/deepspeed/utils/nvtx.py", line 11, in wrapped_fn
      return func(*args, **kwargs)
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/DeepSpeed-master/deepspeed/runtime/engine.py", line 1710, in backward
      self.optimizer.backward(loss)
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/DeepSpeed-master/deepspeed/runtime/zero/stage_1_and_2.py", line 1917, in backward
      self.loss_scaler.backward(loss.float(), retain_graph=retain_graph)
    File "/gpfsssd/worksf/projects/rech/six/commun/code/tr8b-104B/DeepSpeed-master/deepspeed/runtime/fp16/loss_scaler.py", line 53, in backward
      scaled_loss.backward(retain_graph=retain_graph)
    File "/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/_tensor.py", line 363, in backward
      torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
    File "/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/autograd/__init__.py", line 173, in backward
      Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  RuntimeError: CUDA out of memory. Tried to allocate 3.35 GiB (GPU 0; 79.35 GiB total capacity; 70.93 GiB already allocated; 2.20 GiB free; 75.11 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
  
```

thank you!

**Comment by kiukchung** [comment]
Hi @stas00, ha! this is a nice edgecase. I was scratching my head for a while b/c we shouldn't dump unformatted raw errors like this when the "worker" (trainer) fails - that's the whole point of the elastic agent! Filed a separate bug report issue here: https://github.com/pytorch/pytorch/issues/73465 will work on fixing it shortly. Unfortunately we're too late to include this into torch-1.11. In the meanwhile you can work around the annoying error by setting the env var `TORCHELASTIC_ERROR_FILE` :

```
TORCHELASTIC_ERROR_FILE=/tmp/error.json torch.distributed.run ...
```

Longer explanation below:

From
```
 "message": "SignalException: Process 75531 got signal: 15",
```

Was PID 75531 the process that was running `torch.distributed.run` by any chance? I was able to repro this issue by writing a `test.py` that just `time.sleep(forever)`, then running

```
python -m torch.distributed.run <the usual args> test.py
```

And `kill -15`'ing `torch.distributed.run`. The dump is coming from a `log.error` statement here: https://github.com/pytorch/pytorch/blob/df11e2d6f9782fc3995e17ef09a5ef3812da041d/torch/distributed/elastic/multiprocessing/errors/error_handler.py#L36

Basically what is happening is that in the absence of the "error_file"  we dump the raw error to not lose the error (yes to your point we could've formatted it, but it was an exceptional case - shouldn't really have happened, so we just dump it as is). For the workers, the elastic agent always sets the error file (via the `TORCHELASTIC_ERROR_FILE` env var), but in your case nothing was setting the error file of the agent (internally we actually set `TORCHELASTIC_ERROR_FILE` env var to the error file that is picked up by the scheduler when invoking `torch.distributed.run`).

**Comment by stas00** [comment]
Thank you for the feedback and taking care of the edge case, @kiukchung!

If I set `TORCHELASTIC_ERROR_FILE` - I get a single entry logged. Everything else is dumped to the console (383 times)
```
{"message": {"message": "SignalException: Process 17912 got signal: 15", "extraInfo": {"py_callstack": "Traceback (most recent call last):\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py\", line 345, in wrapper\n    return f(*args, **kwargs)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/run.py\", line 724, in main\n    run(args)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/run.py\", line 715, in run\n    elastic_launch(\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/launcher/api.py\", line 131, in __call__\n    return launch_agent(self._config, self._entrypoint, list(args))\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/launcher/api.py\", line 236, in launch_agent\n    result = agent.run()\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/metrics/api.py\", line 125, in wrapper\n    result = f(*args, **kwargs)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/agent/server/api.py\", line 709, in run\n    result = self._invoke_run(role)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/agent/server/api.py\", line 850, in _invoke_run\n    time.sleep(monitor_interval)\n  File \"/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/multiprocessing/api.py\", line 60, in _terminate_process_handler\n    raise SignalException(f\"Process {os.getpid()} got signal: {sigval}\", sigval=sigval)\ntorch.distributed.elastic.multiprocessing.api.SignalException: Process 17912 got signal: 15\n", "timestamp": "1645809651"}}}
```
Is that what's expected?

I'm using `torch.distributed.run` indeed.

> (internally we actually set TORCHELASTIC_ERROR_FILE env var to the error file that is picked up by the scheduler when invoking torch.distributed.run)

I groked all your comments, but got lost on this one. So someone sets `TORCHELASTIC_ERROR_FILE` if I don't already set it? And if so what's the default location?

**Comment by kiukchung** [comment]
> I groked all your comments, but got lost on this one.

There's two levels of error recording. 
1. Worker -> Agent. Each worker writes their errors into an error file. The error filepath is set via `TORCHELASTIC_ERROR_FILE` env var. For the workers, the elastic agent sets this env var, so each agent can collect the structured errors of each of the local workers it manages. This also works for signals (not just exceptions) which in vanilla multiprocessing would not produce such nice exception stack traces.
2. Agent -> Scheduler. There are two scenarios here:
     a. Agent itself raises or signals -> records its own error into the error filepath (also set via `TORCHELASTIC_ERROR_FILE` env var)
     b. Agent observes worker failure -> grabs the "first" (root cause) error observed on the worker and copies the worker error file into the agent's error filepath (`os.environ["TORCHELASTIC_ERROR_FILE]"`)

For our internal use-cases, our schedulers set `TORCHELASTIC_ERROR_FILE` when invoking `python -m torch.distributed.run` so that we can either propagate worker errors or agent failures back to the scheduler.

In your case, no one is setting `TORCHELASTIC_ERROR_FILE` when invoking `python -m torch.distributed.run` hence when the agent errors out (SIGTERM in your case) it wants to record this error, but sees that no one's told it which file to record it at, so it just dumps the structured error (only formats the outer level json, but not the trace field) as an ERROR level log into the console.


> If I set TORCHELASTIC_ERROR_FILE - I get a single entry logged

You mean you see this unformatted json in the file specified by `TORCHELASTIC_ERROR_FILE`? If so yes, this is expected. And yes, the file itself isn't going to be formatted in a human readable way - we could consider making it human readable but I need to check if this will break our error recording/reporting infra, since we internally machine parse these files. If anything I'd have to use different formats for the JSON between OSS and internal, since the current format is the format our internal schedulers understand.

> The rest of the errors are formatted and are dumped to the console (383 times)

Hmmm, did you set `TORCHELASTIC_ERROR_FILE` on all the invocations of `python -m torch.distributed.run`? (and not just on one node). If so you shouldn't see all the console dumps anymore and see something like this (ignore the numpy errors...):

image

**Comment by stas00** [comment]
I did:
```
export TORCHELASTIC_ERROR_FILE=`pwd`/torch-elastic-error.json
python -m torch.distributed.run ...
```

So I'm setting it ahead of time as you suggested.

I get one entry in this file and hundreds of errors dumped to the console.

I additionally added:

```
from torch.distributed.elastic.multiprocessing.errors import record

@record
def main():
     pretrain(train_valid_test_datasets_provider, model_provider, forward_step,
              args_defaults={'tokenizer_type': 'GPT2BPETokenizer'})
```

but I don't see any difference after adding it.

The launcher is:

```
export LAUNCHER="python -u -m torch.distributed.run \
    --nproc_per_node $GPUS_PER_NODE \
    --nnodes $NNODES \
    --rdzv_endpoint $MASTER_ADDR:$MASTER_PORT \
    --rdzv_backend c10d \
    --max_restarts 0 \
    "
```

**Comment by stas00** [comment]
I think there are more issues with `TORCHELASTIC_ERROR_FILE`

I shouldn't need to create this file and the path exists, yes I get this:

```
============================================================
    self._rm(my_error_file)
  File "/gpfswork/rech/six/commun/conda/py38-pt111/lib/python3.8/site-packages/torch/distributed/elastic/multiprocessing/errors/error_handler.py", line 152, in _rm
    os.remove(my_error_file)
FileNotFoundError: [Errno 2] No such file or directory: '/gpfswork/rech/six/commun/code/tr8b-104B/Megatron-DeepSpeed/torch-elastic-error.json'
/usr/bin/bash: line 1: --init-method-std: command not found
/usr/bin/bash: line 1: --init-method-std: command not found
srun: error: jean-zay-iam02: task 1: Exited with exit code 127
srun: Terminating job step 54826.0
srun: error: jean-zay-iam01: task 0: Exited with exit code 127
```

**Comment by kiukchung** [comment]
you’re probably using nfs and setting all the error file locations to the same file. Each node should point to its own file in this case. Typically on schedulers that properly isolate containers this isn’t an issue since the same filepath would actually point to each node’s fs. You could use the /tmp dir of the node (versus a path in the nfs mounted on all the nodes)

**Comment by kiukchung** [comment]
doesn’t matter if you lose access to the error file because the error is re-raised and python will log the exception as usual. It’s a workaround to remove duplicate error logging

**Comment by stas00** [comment]
oh, that's an interesting thought - sure, I'd be glad to try this hack, @kiukchung

**Comment by kiukchung** [comment]
thanks! yeah sorry… i understand it’s an ugly hack for now. Should be resolved for torch-1.12 (unfortunately we missed the torch-1.11 train since we are already on rc and this isn’t a bugfix nor regression)
